Blutooth: Host: Update connection callbacks API documentation and minor fixes

dottspina opened the pull request:
This PR is mostly about updating the API documentation for some connection callbacks: commits titled "Amend xxx() callback documentation", we can also squash them once reviewed.

Additionally, I came across the use of a local variable as an atomic target in `bt_conn_unref()`: either I miss something obvious or it's typo in the source code, may be not deserving a dedicated PR or issue. Since this involves a Zephyr commit that we'll already mention, I've taken the liberty to discuss it briefly here as well.

EDIT: added a trivial commit to fix the MISRA "All if ... else if constructs shall be terminated
with an else clause" (c:M23_112) issues reported by SonarQube and blocking this PR.

## Update connection callbacks API documentation

The API documentation for some connection callbacks appears to not reflect some changes in the Bluetooth Host stack [1] [2] [3].

- `bt_conn_cb.disconnected()`: warns that the connection object is not actually reusable at this stage, but some of the advice given may be deprecated (auto-resume of advertising), and does not mention `recycled()` which IIUC is now the most reliable callback to track free connections and/or initiate operations that depend on them
- `bt_conn_cb.recycled()`: still warns users to "treat this callback as an ISR", although it now runs on the system work-queue thread, as does `disconnected()`
- `bt_conn_cb.connected()`: surprisingly, no special precautions are indicated, while it runs on the same thread that also invokes GATT callbacks such as `bt_gatt_attr.read()` or `bt_gatt_attr.write()` (usually the `BT RX WQ` thread)

The respective commit messages provide some further details.
My wording may not be ideal, edit requests are welcome.

## Use of a local variable as an atomic target    

While reading `subsys/bluetooth/host/conn.c` I came across something that I can't quite grasp, and I'm taking advantage of this PR to address it.

In `bt_conn_unref()`, the local, non shared, variable `old` is used as an atomic *target*:

``` c
atomic_val_t old = atomic_dec(&conn->ref);
conn = NULL;
bool deallocated = (atomic_get(&old) == 1);
```

I can't see what data race or value obsolescence the above call to `atomic_get()` could prevent ?
I considered that this might be a trick to force the compiler to generate some additional memory barrier, but I don't see the need for that either (if so, that might deserve a comment).

If pedantic, one could also note that `atomic_get()` expects an `atomic_t*` argument (called a *target* in Zephyr API), not an `atomic_val_t*` (value).
This compiles without even a warning since Zephyr Atomic defines both to be the same integer type.
The equivalent C11 code, where `_Atomic(T)` and `T` are different types, wouldn't compile.

Either I miss something obvious, or there's some kind of *typo* introduced with [1]. 
Please enlighten me ;-)

Thanks.

[1] efb5d83: Bluetooth: Host: Added Recycled evt notifying conn object is available
[2] 2ca59e7: Bluetooth: Host: Defer `conn.recycled()` to the syswq
[3] 8cfad44: Bluetooth: Deprecate adv auto-resume

dottspina committed:
Bluetooth: Host: Fix use of local variable as atomic target

In bt_conn_unref(), a local variable is used as atomic target:

    atomic_val_t old = atomic_dec(&conn->ref);
    /* Prevent from accessing connection object */
    bool deallocated = (atomic_get(&old) == 1);

Using atomic_get() to access a non-shared local variable
cannot prevent any data race on that variable,
and only causes confusion.

Moreover, this call to atomic_get() is incorrect: the API expects
an atomic_t* argument (target), not an atomic_val_t* (value).
This compiles and /works/ only because Zephyr defines both to be
the same integer type, and thus: atomic_get(&old) == old.
The equivalent C11 code, where _Atomic(T) and T are different types,
wouldn't compile.

Signed-off-by: Christophe Dufaza <chris@openmarl.org>

dottspina committed:
Bluetooth: Host: Amend recycled() callback documentation

The API documentation for the recycled() callback predates [1],
and still warns users to "treat this callback as an ISR",
although it now runs on the system workqueue thread,
as does disconnected().
"Making Bluetooth API calls" to "re-start connectable advertising or
scanning" should no longer be "strongly discouraged".

On the contrary, we can emphasize that this is the right event
to listen for to initiate operations that will try to re-allocate
a freed connection object.

Mention that BT_MAX_CONN configures the size of the connection pool.

Refs:
- [1] efb5d83: Bluetooth: Host: Defer `conn.recycled()` to the syswq

Signed-off-by: Christophe Dufaza <chris@openmarl.org>

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1545,14 +1545,15 @@ void bt_conn_unref(struct bt_conn *conn)
 
 	__ASSERT(old > 0, "Conn reference counter is 0");
 
+	/* Prevent from accessing connection object */
+	conn = NULL;
```
The comment is rather misleading. This is a local variable, and the function doesn't access it anyway after this point.

Thalley replied:
I think the purpose of the comment is specifically to ensure that the local variable cannot accidentally be used. I suppose this is similar to https://doc.rust-lang.org/std/mem/fn.drop.html but not a pattern that we often see in C

dottspina replied:
> The comment is rather misleading. This is a local variable, and the function doesn't access it anyway after this point.

I agree. I kept this comment assuming that the intent was to "Prevent *those who will read this in ten years* from accessing connection object *from here until the end of the function*".

IMO we should either:
- move a reworded comment and the `conn = NULL` statement to the `if (deallocated)` block (disposing the connection pointer regardless of its reference counter is also confusing)
- or remove both the comment and the `conn = NULL` statement

dottspina commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1660,6 +1660,11 @@ struct bt_conn_cb {
 	 *        callback. Instead use the connected callback of the
 	 *        advertising set.
 	 *
+	 *  @note This callback will run on the same, non-preemptible, work-queue thread
+	 *        that processes incoming low priority HCI packets and invokes GATT callbacks
+	 *        such as @ref bt_gatt_attr.read and @ref bt_gatt_attr.write.
+	 *        Blocking operations are therefore discouraged.
```
> I'm working on moving all host-internal processing to a single internal work queue [...] while at the same time deferring any application callbacks to the system workqueue. 

This will indeed be a welcome improvement:  we'd better wait until you have made progress with this refactoring before updating the related API documentation (is there already an RFC or some draft ?).

jhedberg replied:
> (is there already an RFC or some draft ?)

Not yet. I have a branch in my Zephyr fork, but it's still pretty messy. I'll create a draft PR once it's a bit cleaned up. Also, I don't think I can do all the wanted changes in one go. First step is to create the generic internal workqueue and move as much as possible of internal processing to it. Moving *all* callbacks then to the system workqueue will need to be a separate (later) step, since that will require some more intrusive changes.

dottspina committed:
Bluetooth: Host: Fix some MISRA c:M23_112 warnings

Fix some "All if ... else if constructs shall be terminated
with an else clause" (c:M23_112) issues reported by SonarQube.

Signed-off-by: Christophe Dufaza <chris@openmarl.org>

dottspina commented:
Sorry @Thalley , our messages from Friday crossed.
I think we all agree on preventing access to a possibly recycled connection object, I just reworded the comment which I too found misleading: either it is unnecessary, assuming that the purpose of the `conn = NULL;` statement is obvious, or we should be careful not to suggest that this statement is, in some opaque way, about the actual connection object.
 
 @jhedberg , understood, move forward quietly, we keep my proposals for the moment, and we'll review the documentation of all related callbacks when you're ready.

I also added a trivial commit to fix the MISRA "All if ... else if constructs shall be terminated
with an else clause" (c:M23_112) issues reported by SonarQube and blocking this PR.

Thanks.

dottspina commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1690,9 +1690,9 @@ struct bt_conn_cb {
-	 *  To avoid this issue it is recommended to either start connectable
-	 *  advertise or create a new connection using @ref k_work_submit or
-	 *  increase @kconfig{CONFIG_BT_MAX_CONN}.
+	 *  To avoid this issue it is recommended to either use @ref k_work_submit
+	 *  or rely on @ref bt_conn_cb.recycled to create new connections
+	 *  or resume connectable advertising.
```
These are actually recommendations, not suggestions, which is why I thought the documentation was out of date.

I agree that setting `BT_MAX_CONN` to N+1 when planning N simultaneous connections is a work-around that users may have become accustomed to. And that was probably the easiest approach for some time, rightly documented as such. 

But should we continue to recommend it?
IMO, we should stop documenting how to create new connections or restart advertising from the `disconnected()` callback, and instead simply recommend relying on `recycled()` for these use cases. No work-around needed.

I even considered removing `k_work_submit()` from the recommendations as well: I didn't go that far, but it could make sense.

Tell me what you think, and I'll update the documentation.

I can also reword the commit message: would it be okay if I presented "increase `BT_MAX_CONN`" as a deprecated work-around ?

alwa-nordic replied:
I think we should remove the recommendation the way it is worded currently. Simply scheduling an advertiser start using `k_work_submit()` is not a reliable pattern.

We can definitely direct users towards `recycled()` instead. This is the right event to listen for to start a new connection or advertiser both when the intention is to start it after the system is completely finished with an earlier connection, and when the application wants start a connection for any reason but failed and is waiting for the right time to retry.

If the application wants to start a new connection or advertiser after a specific connection is gone, then it should note the when that connection becomes `disconnected()` and then keep trying to start the new connection or advertiser on `recycled()`.

In any case, if on a `recycled()` event, creating a connection or starting an advertiser fails, then the application should retry on the next `recycled()` event. Note here the `bt_conn_le_create()` "Found valid connection" problem.

dottspina replied:
@alwa-nordic , thanks for confirming this.
I liked your explanatory approach, and added it to the API documentation.

dottspina commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1701,10 +1706,8 @@ struct bt_conn_cb {
-	 * @ref bt_conn_unref which is used by the BT stack. Making
-	 * Bluetooth API calls in this context is error-prone and strongly
-	 * discouraged.
+	 * This callback runs on the system-worqueue thread,
+	 * the usual precautions apply.
```
Absolutely right.

However:
- today `recycled()` runs on the system work-queue: knowing this can help, e.g. to avoid inversion of priority issues
- AFAIK @jhedberg is working on deferring all application callbacks to the system work-queue

> Not 100% opposed to this, but not sure it should be added either

Not sure either.
My two cents: it's worth having documentation that's faithful to current behavior, to better guide users, and it will make it easier to update next time after @jhedberg's work is complete.

dottspina commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1530,40 +1530,49 @@ static K_WORK_DEFINE(recycled_work, recycled_work_handler);
+		 * This behavior is deprecated:
+		 * - 8cfad44: Bluetooth: Deprecate adv auto-resume
+		 * - #72567: Bluetooth: Advertising resume functionality is broken
+		 * - Migration guide to Zephyr v4.0.0, Automatic advertiser resumption is deprecated
```
In addition to being deprecated, it seems to me that this behavior is anyway rather strange:
1. deallocate the connection
2. submit a work that, when processed, will notify listeners that a connection object has been freed and can be taken
3. immediately reuse this freed slot to resume advertising

When the work is processed and `recycled()` runs, the listeners may incorrectly assume that at least one connection object is available.

Thalley replied:
Yeah, that's not great :) But we should be able to remove the auto-resume completely now, so that should clear up that mess :D

Thalley replied:
> Are there other places where the Bluetooth stack can internally take a connection object without the application itself calling the public API, e.g. bt_conn_le_create() or bt_le_adv_start() ?

No, but the `recycled` callback is called in a loop

```c
	STRUCT_SECTION_FOREACH(bt_conn_cb, cb) {
		if (cb->recycled) {
			cb->recycled();
		}
	}
```

So if the owner of the first registered callback allocates a new connection object (e.g. by restarting advertisement), then the owner of the second registered callback won't be able to do that

dottspina replied:
Agreed, "No guarantee, first come, first serve" actually makes sense as soon as there is more than one registered listener. Keeping this statement as-is in the API documentation.

dottspina committed:
Bluetooth: Host: Amend disconnected() callback documentation

The API documentation for the disconnected() callback warns that
the listener can't assume that the corresponding connection object
has been freed and may me available to the application.

The recommendations given to still start a new connection
or connectable advertiser are outdated or misleading:
- "start connectable advertising": the options that "will attempt
  to resume the advertiser under some conditions" are deprecated
  since Zephyr 4.0 (BT_LE_ADV_OPT_CONNECTABLE and related)
- "using k_work_submit()": assuming everything will be fine
  when the work is actually processed is not reliable
- "increase CONFIG_BT_MAX_CONN": setting BT_MAX_CONN to N+1
  when planning N simultaneous connections is a work-around
  that users may have gotten used to (despite its footprint),
  but there is no longer any reason to advise it

Stop documenting creating new connections or restarting advertising
from the disconnected() callback and instead recommend relying
on recycled() for these use cases.

Signed-off-by: Christophe Dufaza <chris@openmarl.org>

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1849,12 +1856,16 @@ struct bt_conn_cb {
 	 * allocate a connection object. No guarantee, first come, first serve.
 	 *
-	 * Use this to e.g. re-start connectable advertising or scanning.
+	 * The maximum number of simultaneous connections is configured
+	 * by @kconfig{CONFIG_BT_MAX_CONN}, which usually defaults to a single connection.
```
```suggestion
	 * The maximum number of simultaneous connections is configured
	 * by @kconfig{CONFIG_BT_MAX_CONN}.
```

I don't think the second part of this sentence is helpful or necessary. I'd rather just refer to the Kconfig option's default value

dottspina replied:
Okay, I'll remove this second part.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1849,12 +1856,16 @@ struct bt_conn_cb {
+	 * The maximum number of simultaneous connections is configured
+	 * by @kconfig{CONFIG_BT_MAX_CONN}, which usually defaults to a single connection.
+	 *
+	 * This is the event to listen for to start a new connection, connectable advertiser
+	 * or scanner, both when the intention is to start it after the system is completely
```
```suggestion
	 * This is the event to listen for to start a new connection or connectable advertiser,
	 * both when the intention is to start it after the system is completely
```

Initiating scanning is not directly related to creating a new connection object, and restarting scanning in this, or the disconnected callback, should always work

dottspina replied:
Thank you for clarifying this, I wasn't entirely convinced myself.

This reference to the scanner comes from the existing API documentation:
> Use this to e.g. re-start connectable advertising or scanning.

which I may have misinterpreted.
I'll fix the API doc.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1849,12 +1856,16 @@ struct bt_conn_cb {
-	 * @ref bt_conn_unref which is used by the BT stack. Making
-	 * Bluetooth API calls in this context is error-prone and strongly
-	 * discouraged.
+	 * This callback runs on the system workqueue thread,
+	 * the usual precautions apply.
```
What are the usual precautions? I think that if we want to write such as statement, we should define those and refer to it. If you don't know the usual precautions, then this isn't helpful :)

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1492,33 +1492,46 @@ void bt_conn_unref(struct bt_conn *conn)
 
-	/* Slot has been freed and can be taken. No guarantees are made on requests
-	 * to claim connection object as only the first claim will be served.
+	deallocated = (old == 1);
```
```suggestion
	deallocated = (atomic_get(&old) == 1);
```

Given that `old` is a `atomic_val_t`, we should use the appropriate getter for it

Thalley replied:
No, I think you are right here. In fact, `atomic_get` returns an `atomic_val_t`, so my suggestion was fundamentally invalid :D

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2399,6 +2416,8 @@ struct bt_conn *bt_conn_add_sco(const bt_addr_t *peer, int link_type)
 	} else if (link_type == BT_HCI_ESCO) {
 		sco_conn->sco.pkt_type = (bt_dev.br.esco_pkt_type &
 					  ~EDR_ESCO_PKT_MASK);
+	} else {
+		/* Ignoring unexpected link type BT_HCI_ACL. */
```
I assume that this was added due to Sonarqube - Since it's just a comment it's OK, but otherwise just ignore it next time that Sonar complains about code you haven't modified.

dottspina replied:
Yes, this was to fix a Sonarqube warning (and there's another similar change in the same file).
Ok, from now on I'll ignore it when it complains about code I didn't change, thanks.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1835,9 +1839,12 @@ struct bt_conn_cb {
+	 *  a connection object has actually been freed.
+	 *
+	 * This callback runs on the system workqueue thread,
+	 * the usual precautions apply.
```
> * the test can't happen within the `__ASSERT()` statement: when assertions are disabled, the underlying variable will be unused, possibly causing build failures depending on compiler flags (see CI failures); I couldn't figure out how to fix that elegantly, or at least without adding even more ugliness (e.g. a dummy statement _using_ the _unused_ variable outside of the assertion)

Hmm, it's odd that we see a CI issue here because ASSERT should be enabled in those, and even then we have many other cases where local variables are used only for __ASSERT without causing CI issues (sonarqube and coverity sometimes complain about those though). 

Some not-very-nice-looking solutions could be to modify the `IF_ENABLED(CONFIG_BT_CONN_TX` to also check `CONFIG_ASSERT` or to apply `__maybe_unused` to the local variable.

dottspina committed:
Bluetooth: Host: Fix possible inconsistent access to connection state

An assertion in bt_conn_unref() accesses the connection's state
after decrementing its reference count.
This is not consistent since, if we removed the last reference,
the Bluetooth Host stack may reuse the connection object
before the assertion is checked.
Instead, retrieve the connection property tested by the assertion
before decrementing the counter, as we do for other properties.

Simplify the code path by returning early when we did not remove
the last reference.

Remind that automatic advertiser resumption is deprecated.

Signed-off-by: Christophe Dufaza <chris@openmarl.org>

dottspina commented:
@Thalley, thank you for your review.

This PR should now only contain changes we agree on:
- In `bt_conn_unref()` (2 commits): fix use of local variable as atomic target, fix connection usage after decrementing its reference count
- API documentation update (2 commits): clearly state that `recycled()` is now the event to listen for to start a new connection or connectable advertiser, do not recommend workarounds for  `disconnected()` (confirmed by @alwa-nordic)

Other considerations that emerged from our discussion:
- It is useful to know the execution context of a callback that one implements, and the API documentation could help with this. Perhaps it's enough to tell users which thread, and assume they do their homework to understand the implications. What to do and how (e.g. Doxygen aliases ?) is beyond the scope of this PR and may deserve an RFC (e.g. starting with the Bluetooth subsystem).
- Automatic advertiser resumption is deprecated since Zephyr 4.0: I considered removing the relevant part from `bt_conn_unref()`, but it wouldn't be consistent if we didn't  remove its support altogether (configuration options and related API). I simply added a comment so we don't forget,  but it might be worth opening an RFC (AFAICT samples and tests have already been updated).

Thanks.

Thalley reviewed (approved):
LGTM

If you want to follow up with another PR, I noticed that our ADV API does not mentioned that it reserves a `bt_conn` object if starting connectable advertising, and that it is also not mentioned that if you stop connectable advertising, then the `recycled` callback is likewise called :)
